src/manifest/pfm.rs

```diff
@@ -412,88 +448,120 @@ where
                 let mut hash = [0; 32];
                 sha.hash_contiguous(data, &mut hash)?;
                 if &hash != expected {
-                    return Err(Error::SignatureFailure);
+                    return Err(Error::BadElementHash {
+                        toc_index: self.entry.index(),
+                    });
                 }
             }
         }
 
-        let mut buf = data;
-        let image_count = buf.read_le::<u8>()? as usize;
-        let rw_count = buf.read_le::<u8>()? as usize;
-        let version_len = buf.read_le::<u8>()? as usize;
-        let _ = buf.read_le::<u8>()?;
-        let version_addr = buf.read_le::<u32>()?;
-
-        if buf.len() < version_len {
-            return Err(Error::OutOfRange);
+        #[derive(FromBytes)]
+        #[repr(C)]
+        struct FwVersionHeader {
+            image_count: u8,
+            rw_count: u8,
+            version_len: u8,
+            _unused: u8,
+            version_addr: u32,
+        }
+        let (header, rest) =
+            LayoutVerified::<_, FwVersionHeader>::new_from_prefix(data).ok_or(
+                Error::TooShort {
+                    toc_index: self.entry.index(),
+                },
+            )?;
+
+        if rest.len() < header.version_len as usize {
+            return Err(Error::TooShort {
+                toc_index: self.entry.index(),
+            });
         }
-        let (version_str, mut buf) = buf.split_at(version_len);
+        let (version_str, mut buf) = rest.split_at(header.version_len as usize);
 
         // Align back to 4-byte boundary.
-        buf = &buf[misalign_of(buf.as_ptr() as usize, 4)..];
+        buf = buf.get(misalign_of(buf.as_ptr() as usize, 4)..).ok_or(
+            Error::TooShort {
+                toc_index: self.entry.index(),
+            },
+        )?;
 
-        let rw_len = mem::size_of::<RwRegion>() * rw_count;
+        let rw_len = mem::size_of::<RwRegion>() * header.rw_count as usize;
+        if buf.len() < rw_len {
+            return Err(Error::TooShort {
+                toc_index: self.entry.index(),
+            });
+        }
         let (rw_bytes, unparsed_image_regions) = buf.split_at(rw_len);
         // NOTE: This cannot panic, since it checks for alignment (4-byte) and
         // size, which have already been explicitly checked above.
         let rw_regions = LayoutVerified::<_, [RwRegion]>::new_slice(rw_bytes)
             .unwrap()
             .into_slice();
+
         for rw in rw_regions {
-            if rw.start_addr > rw.end_addr {
-                return Err(Error::OutOfRange);
-            }
+            Region::from_start_and_limit(rw.start_addr, rw.end_addr).ok_or(
```

Comment:
Creates and discards a new Region just for the error checking properties of `from_start_and_limit`?

More clear as `let _ = Region::from_start_and_limit(...).ok_or(...)?;`

(Not sure on what is idiomatic Rust in this case).

Reply:
Yup, that makes sense. I don't think the fact that it was called just for error-checking is obvious.